### PR TITLE
fix: when missing _X_AMZN_TRACE_ID skip tracer

### PR DESCRIPTION
**Pull request by OrrLumigo**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -256,7 +256,8 @@ export const isSwitchedOff = (): boolean =>
     () =>
       validateEnvVar(SWITCH_OFF_FLAG) ||
       TracerGlobals.getTracerInputs().switchOff ||
-      !isValidAlias()
+      !isValidAlias() ||
+      !process.env['_X_AMZN_TRACE_ID']
```
Comment (saartochner) [inline src/utils.ts:260]: change to `isAwsEnvironment`

File: src/tracer.ts [hunk 2]
```diff
@@ -224,12 +224,16 @@ export const trace =
       return performPromisifyType(err, data, type, callback);
     }
     context.__wrappedByLumigo = true;
+    let pStartTrace = Promise.resolve();
+    let functionSpan;
+    if (!isSwitchedOff()) {
```
Comment (saartochner) [inline src/tracer.ts:229]: I think that this `if` should be the first line in this function (`trace`), and it should immediately stop everything (all the code lines in here should not run!)

File: src/tracer.test.js [hunk 3]
```diff
@@ -482,6 +482,23 @@ describe('tracer', () => {
     expect(Http.hookHttp).toHaveBeenCalledTimes(1);
   });
 
+  test('trace; async local lambda no aws _X_AMZN_TRACE_ID', async () => {
+    delete process.env['_X_AMZN_TRACE_ID'];
```
Comment (saartochner) [inline src/tracer.test.js:486]: This line is not mandatory (because of the next line), right?

File: src/tracer.test.js [hunk 4]
```diff
@@ -482,6 +482,23 @@ describe('tracer', () => {
     expect(Http.hookHttp).toHaveBeenCalledTimes(1);
   });
 
+  test('trace; async local lambda no aws _X_AMZN_TRACE_ID', async () => {
```
Comment (saartochner) [inline src/tracer.test.js:485]:
```suggestion
  test('trace; async local lambda no aws environment', async () => {
```

File: src/tracer.ts [hunk 5]
```diff
@@ -197,6 +197,13 @@ export const trace =
   ({ token, debug, edgeHost, switchOff, stepFunction }) =>
   (userHandler) =>
   async (event, context, callback) => {
+    if (!isAwsEnvironment()) {
+      let handlerReturnValue = await promisifyUserHandler(userHandler, event, context);
+      handlerReturnValue = normalizeLambdaError(handlerReturnValue);
+      const cleanedHandlerReturnValue = removeLumigoFromStacktrace(handlerReturnValue);
+      const { err, data, type } = cleanedHandlerReturnValue;
+      return performPromisifyType(err, data, type, callback);
```
Comment (saartochner) [inline src/tracer.ts:205]: why not simply:
`return userHandler(event, context)`
?

File: src/utils.ts [hunk 6]
```diff
@@ -169,7 +169,12 @@ const LUMIGO_STACK_PATTERNS = [
 const validateEnvVar = (envVar: string, value: string = 'TRUE'): boolean =>
   !!(process.env[envVar] && process.env[envVar].toUpperCase() === value.toUpperCase());
 
-export const isAwsEnvironment = () => !!process.env['LAMBDA_RUNTIME_DIR'];
+export const isAwsEnvironment = () =>
+  !!(
+    process.env['LAMBDA_RUNTIME_DIR'] &&
+    !process.env['AWS_SAM_LOCAL'] && // local SAM
+    !process.env['IS_LOCAL']
+  ); // local SLS
```
Comment (saartochner) [inline src/utils.ts:177]:
```suggestion
    !process.env['IS_LOCAL'] // local SLS
  ); 
```

File: src/tracer.test.js (hunk 3 above)
Comment (OrrLumigo) [reply src/tracer.test.js:486]: No i want see that its not crushing when its missing

File: src/tracer.ts (hunk 5 above)
Comment (OrrLumigo) [reply src/tracer.ts:205]: what if it comes with callback?

File: src/utils.ts (hunk 6 above)
Comment (OrrLumigo) [reply src/utils.ts:177]: lint auto doing it

File: src/tracer.ts (hunk 5 above)
Comment (OrrLumigo) [reply src/tracer.ts:205]: and if there is exception we need to remove lumigo from there

Comment (saartochner) [reply src/tracer.ts:205]: of course, add the callback.
I don't want to run any logic here. Even not the remove-from-stacktrace IMO

File: src/tracer.test.js (hunk 3 above)
Comment (nirLumigo) [reply src/tracer.test.js:486]: Maybe we can do `parametrize`?  One with "_X_AMZN_TRACE_ID" and one without "_X_AMZN_TRACE_ID"?

**Review by nirLumigo — APPROVED** [review]

**Review by saartochner — APPROVED** [review]

**Pull request merged**